operator/cmd/k8s_identity_gc_test.go

```diff
@@ -0,0 +1,199 @@
+	goleak.VerifyTestMain(
+		m,
+		goleak.Cleanup(
+			func(exitCode int) {
+				// Force garbage-collection to force finalizers to run and catch
+				// missing Event.Done() calls.
+				runtime.GC()
+			}),
+		// To ignore goroutine started from sigs.k8s.io/controller-runtime/pkg/log.go
+		// init function
+		goleak.IgnoreTopFunction("time.Sleep"),
+
+		// To ignore leaked goroutine started from the instantiation
+		// of global variable watchers.nodeQueue
+		goleak.IgnoreTopFunction("k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop"),
+	)
+}
+
+func TestIdentitiesGC(t *testing.T) {
+	clientset, _ := k8sClient.NewFakeClientset()
+	nodes := []*corev1.Node{
+		{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Node",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "identity-control-plane",
+				Labels: map[string]string{"kubernetes.io/hostname": "identity-control-plane"},
+			},
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Node",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "identity-worker",
+				Labels: map[string]string{"kubernetes.io/hostname": "identity-worker"},
+			},
+		},
+	}
+	for _, node := range nodes {
+		if _, err := clientset.CoreV1().Nodes().
+			Create(context.Background(), node, metav1.CreateOptions{}); err != nil {
+			t.Fatalf("Failed to create node %v: %s", node, err)
+		}
+	}
+
+	identity := &v2.CiliumIdentity{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "cilium.io/v2",
+			Kind:       "CiliumIdentity",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "99999",
+			Labels: map[string]string{
+				"io.cilium.k8s.policy.cluster":        "default",
+				"io.cilium.k8s.policy.serviceaccount": "default",
+				"io.kubernetes.pod.namespace":         "default",
+				"test":                                "identities-gc",
+			},
+		},
+	}
+	if _, err := clientset.CiliumV2().CiliumIdentities().
+		Create(context.Background(), identity, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create identity %v: %s", identity, err)
+	}
+
+	hive := hive.New(
+		// provide a mock clientset
+		cell.Provide(func() k8sClient.Clientset {
+			return clientset
+		}),
+
+		// provide a identities resource events
+		cell.Provide(func(lc hive.Lifecycle, c k8sClient.Clientset) resource.Resource[*v2.CiliumIdentity] {
+			return resource.New[*v2.CiliumIdentity](
+				lc, utils.ListerWatcherFromTyped[*v2.CiliumIdentityList](
+					c.CiliumV2().CiliumIdentities(),
+				),
+			)
+		}),
+
+		cell.Provide(newIdentitiesGC),
+		cell.Provide(newTestIdentitiesGCConfig),
+
+		// force instantiation of identity-gc to start the test
+		cell.Invoke(func(*IdentitiesGC) {}),
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if err := hive.Start(ctx); err != nil {
+		t.Fatalf("Failed to start: %s", err)
+	}
+
+	// signal that endpoints are sync-ed, otherwise identities gc won't start
+	close(watchers.CiliumEndpointsSynced)
```

Comment:
Should add `Resource[*CiliumEndpoints]` and get rid of this. The test also doesn't seem to check whether we correctly keep identities that have a CiliumEndpoint associated with them.

Reply:
We'll also need to have the CiliumEndpoints indexed by identity ID (e.g. with something like: https://github.com/joamaki/cilium/blob/pr/joamaki/operator-resource-trials/pkg/k8s/resource/index.go).

Reply:
Yep, I'll make this the priority for the next architecture-related PR.